HealthcareApp/src/Healthcare.Infrastructure/Repository/AppointmentRepository.cs

```diff
@@ -46,12 +54,39 @@ public async Task<bool> IsAvailableAsync(string doctorId, DateTime date, TimeSpa
     {
         var endTime = startTime.Add(TimeSpan.FromMinutes(15));
 
-        var existingAppointment = await _appDbContext.Appointments.SingleOrDefaultAsync( a =>
+        var existingAppointment = await _appDbContext.Appointments.SingleOrDefaultAsync(a =>
             a.DoctorId == doctorId &&
             a.AppointmentDate == date &&
             ((a.StartTime <= startTime && a.EndTime > startTime) ||
              (a.StartTime < endTime && a.EndTime >= endTime)));
 
         return existingAppointment == null;
     }
+
+    public async Task UpsertAppointmentAsync(Appointment appointment)
+    {
+        using var sqlConnection = _dbConnectionFactory.CreateDbConnection();
+
+        const string query = """
+                                         MERGE INTO Appointments AS target
+                                         USING (SELECT @Id AS Id) AS source
+                                         ON (target.Id = source.Id)
+                                         WHEN MATCHED THEN 
+                                             UPDATE SET 
+                                                 DoctorId = @DoctorId,
+                                                 PatientId = @PatientId,
+                                                 AppointmentDate = @AppointmentDate,
+                                                 StartTime = @StartTime,
+                                                 EndTime = @EndTime,
+                                                 Status = @Status,
+                                                 DurationMinutes = @DurationMinutes
+                                         WHEN NOT MATCHED THEN
+                                             INSERT (DoctorId, PatientId, AppointmentDate, StartTime, EndTime, Status, DurationMinutes)
+                                             VALUES (@DoctorId, @PatientId, @AppointmentDate, @StartTime, @EndTime, @Status, @DurationMinutes);
+                             """;
+
+
+        appointment.DurationMinutes = (int)(appointment.EndTime - appointment.StartTime).TotalMinutes;
```

Comment:
Should be moved to Application layer because it's not related to data storage concerns